File: src/codes/scattered-instaniation/py/rfc_builder.py
class Logger:
    def log(self, message: str):
        print(f"[LOG]: {message}")

class UserService:
    def __init__(self, logger: Logger):
        self.logger = logger

    def create_user(self, name: str):
        self.logger.log(f"ユーザー作成: {name}")

class UserServiceBuilder:
    def __init__(self):
        self._logger: Logger | None = None

    def set_logger(self, logger: Logger) -> "UserServiceBuilder":
        self._logger = logger
        return self

    def build(self) -> UserService:
        if not self._logger:
            raise ValueError("Logger is required")
        return UserService(self._logger)

# 利用例
builder = UserServiceBuilder()
service = builder.set_logger(Logger()).build()
service.create_user("Hiroshi")

File: src/codes/scattered-instaniation/php/rfc_builder.php
<?php

class Logger {
  public function log(string $message): void {
    echo "[LOG]: {$message}\n";
  }
}

class UserService {
  public function __construct(private Logger $logger) {}

  public function createUser(string $name): void {
    $this->logger->log("ユーザー作成: {$name}");
  }
}

class UserServiceBuilder {
  private ?Logger $logger = null;

  public function setLogger(Logger $logger): self {
    $this->logger = $logger;
    return $this;
  }

  public function build(): UserService {
    if (!$this->logger) {
      throw new Exception("Logger is required");
    }
    return new UserService($this->logger);
  }
}

// 利用例
$builder = new UserServiceBuilder();
$service = $builder->setLogger(new Logger())->build();
$service->createUser("Hiroshi");
